File: packages/shared-types/src/provider-calls.ts
/** Mirrors `app/schemas/provider_calls.py`. Audit trail of engine/provider calls per analysis. */

export type ProviderCallStatus = "success" | "cached" | "failed" | "timeout" | "skipped";

export interface ProviderCall {
  id: string;
  provider: string;
  operation: string;
  model_version: string | null;
  request_id: string | null;
  status: ProviderCallStatus;
  latency_ms: number | null;
  estimated_cost: number | null;
  /** SHA-256 of the request content; never the content itself. */
  request_hash: string | null;
  response_json: Record<string, unknown> | null;
  error_json: Record<string, unknown> | null;
  created_at: string;
}

export interface ProviderCallsResponse {
  calls: ProviderCall[];
  total_estimated_cost: number;
  currency: string;
}

File: services/api/app/schemas/provider_calls.py
import uuid
from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict

from app.enums import ProviderCallStatus


class ProviderCallResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    provider: str
    operation: str
    model_version: str | None
    request_id: str | None
    status: ProviderCallStatus
    latency_ms: int | None
    estimated_cost: float | None
    request_hash: str | None
    response_json: dict[str, Any] | None
    error_json: dict[str, Any] | None
    created_at: datetime


class ProviderCallsResponse(BaseModel):
    calls: list[ProviderCallResponse]
    total_estimated_cost: float
    currency: str = "USD"
